### DISCUSSION TITLE
How to make mikro-orm work with esbuild?

### DISCUSSION THREAD

**Discussion by standuprey** [mikro-orm/mikro-orm#2219] [Q&A]
**Is your feature request related to a problem? Please describe.**
I am exploring some ways to optimize our project build time. esbuild is a lot faster than tsc. But it doesn't produce the same output as tsc, simply a .js bundled file

**Describe the solution you'd like**
A way to discover entities in this situation, I know that there is a Javascript metadata provider, but it is deprecated , so I am hesitant to use it.

**Describe alternatives you've considered**
only use tsc to generate .d.ts files for the entities, but it makes my build process unnecessary complex

**Comment by B4nan** [comment]
Can't help with esbuild, but if you want to have a single JS bundle, it sounds pretty much the same as webpack, so the same restrictions will apply. 

https://mikro-orm.io/docs/deployment/#deploy-a-bundle-of-entities-and-dependencies-with-webpack

Namely you can't use folder based discovery, and basically anything that would result in dynamic requires/imports.

Looking at slack history, I see @thomaschaaf managed to [make it work somehow](https://mikroorm.slack.com/archives/CFTFEK39D/p1615068253196800). You will probably need his https://github.com/thomaschaaf/esbuild-plugin-tsc. Would be great to have a recipe about that in the docs.

> A way to discover entities in this situation, I know that there is a Javascript metadata provider, but it is deprecated , so I am hesitant to use it.

It is deprecated because you should use [`EntitySchema`](https://mikro-orm.io/docs/entity-schema/) instead, which is metadata provider independent.

**Reply by standuprey** [reply]
thanks for the pointer, will look into this and make a write up if I'm successful

**Comment by sadtomatoofjoy** [comment]
Hi I've managed to setup mikroorm with esbuild and serverless and just wanted to share it here.

here is the config:

```
orm = await MikroORM.init({
      host: process.env.DB_HOST,
      user: process.env.DB_USER,
      password: process.env.DB_PASSWORD,
      dbName: process.env.DB_NAME,
      port: Number(process.env.DB_PORT),
      type: 'postgresql',
      entities: entities,
      debug: true,
      implicitTransactions: false,
      loadStrategy: LoadStrategy.JOINED,
      discovery: { warnWhenNoEntities: false },
    }
```

So i pass the list of entities instead of using paths. 
The metadata provider is reflect metadata.

With webpack, i didn't have to provide the type option in the @Property decorator in my entities but i had to define it with esbuild.

**edit**: it's because esbuild doesn't support metadata so you either have to add a metadata plugin or try not to use them (like in this case).

Lastly, this error occured: https://github.com/evanw/esbuild/issues/473#issue-724783093.
but it wasn't possible to just pass the client to knex as advised in the issue because it was hardcoded in the @mikro-orm/postgresql/PostgreSqlConnection.js class (we are using pg in our project).
```
    async connect() {
        this.patchKnex();
        this.client = this.createKnexClient('pg');
    } 
```

So i had to redefine this function globally:
```
import { PostgreSqlConnection } from '@mikro-orm/postgresql';
import KnexPostgres from 'knex/lib/dialects/postgres';

PostgreSqlConnection.prototype.connect = async function (this: PostgreSqlConnection) {
  this['patchKnex']();
  this.client = this.createKnexClient(KnexPostgres);
};
```
and inject it via the inject option of esbuild.

The esbuild configuration used 
```
bundle: true
minify: true
packager: yarn
exclude: [ 'aws-sdk', 'pg-native', '@mikro-orm/entity-generator',  # all the drivers and libraries that are not used
           '@mikro-orm/migrations', '@mikro-orm/mongodb', '@mikro-orm/mysql',
           '@mikro-orm/mariadb', '@mikro-orm/sqlite', 'pg-query-stream', 'sqlite3',
           'mysql' ]
inject: [ '../../esbuild-mikroorm-patch.ts' ] # this is the code mentionned above
```

**Reply by Compulsed** [reply]
I got it working, thanks for sharing! I'm using the CDK so I had to follow these steps. If anyone finds any way to improve on the above please share.

**Reply by SamHatoum** [reply]
@Compulsed could you kindly share your config for how you got the CDK to work?

**Reply by B4nan** [reply]
Note that v6 removes those problematic require statements (#3814), so the `exclude` block won't be needed anymore.

**Reply by SamHatoum** [reply]
Thanks @B4nan, that's great news. ~I just tried v6 and received a `TypeError: ext is not a function`, so probably best to wait for a more stable version I guess! (happy to provide more feedback if you need it)~ 
EDIT: turns out I didn't also upgrade the mysql driver

**Comment by parisholley** [comment]
i use https://www.npmjs.com/package/esbuild-plugin-decorator with compiler set to `swc` for my remix project so that decorators work. be sure to set `keepNames: true` in your esbuild options. also use a convention to "register" all of my entities like so:

```typescript
export const mikros = new Map<string, EntityClass<AnyEntity>>();

export function registerMikro(klazz: EntityClass<AnyEntity>) {
  mikros.set(klazz.name, klazz);
}

@Entity()
class MyEntity {
}

registerMikro(MyEntity);

MikroORM.init<PostgreSqlDriver>({
    entities: Array.from(mikros.values())
})
```

**Reply by SamHatoum** [reply]
Thanks! I'll give this a whirl
